### PR TITLE
feature/settings-username-update

### PR DESCRIPTION
**Pull request by alannnc**
To be merged with calcom/website#136

## What does this PR do?

Implements username availability upgrade/downgrade from settings/profile

## Type of change

- [X] New feature (non-breaking change which adds functionality)

## How should this be tested?

Recording:
https://www.loom.com/share/153ab1e374a2455bab2bfc0aa1f14e00

- [ ] Stripe credentials, Stripe Product Ids and Stripe Prices Ids are required for testing
- [ ] A user with subscriptionId from stripe
- [ ] Use a Free/Trial user and try to upgrade your username to a premium one, should navigate to checkout stripe.
- [ ] Use a Pro user and try to upgrade from a standard username to a premium one, should navigate to stripe update subscription page.
- [ ] Use a Pro user with a premium username and downgrade to a standard username, should navigate to stripe update subscription page.

### CODE REVIEW

File: apps/web/pages/api/username/index.ts
Comment (zomars) [inline apps/web/pages/api/username/index.ts:4]: We cannot leave this like this. This will cause errors on self-hosted users since they don't have this module. Let's just stick to calling the `cal.com/api/username` endpoint directly from here.

**Review by zomars — CHANGES_REQUESTED** [review]

File: packages/prisma/migrations/20220316055019_added_username_premium_boolean_to_user_model/migration.sql
Comment (zomars) [inline packages/prisma/migrations/20220316055019_added_username_premium_boolean_to_user_model/migration.sql:2]: Let's put this inside the metadata json field since it's not a core feature for product.

File: apps/web/pages/api/username/index.ts
Comment (alannnc) [reply apps/web/pages/api/username/index.ts:4]: Yeah this change was before you helped me the other day with the same issue. Would keep this as a reminder to use api endpoint

File: apps/web/pages/settings/profile.tsx
Comment (alannnc) [inline apps/web/pages/settings/profile.tsx:105]: Should this be mode: cors?

File: .env.example
Comment (alannnc) [inline .env.example:87]: Now we fetch subscription and try to match its product id vs this env var

File: apps/web/components/ui/UsernameAvailability.tsx
Comment (alannnc) [inline apps/web/components/ui/UsernameAvailability.tsx:281]: added index="desktop" so I can search properly while running tests and don't run into dup data-testids

File: packages/app-store/stripepayment/api/subscription.ts
Comment (alannnc) [inline packages/app-store/stripepayment/api/subscription.ts:85]: Here we tell stripe to make a proration on current subscription, this means charge will happen after billing cycle reach its due date. Or should we immediately charge? @zomars

File: packages/prisma/migrations/20220513213939_add_intent_username_table/migration.sql
Comment (alannnc) [inline packages/prisma/migrations/20220513213939_add_intent_username_table/migration.sql:2]: This table saves all intent username that user wants before moving to stripe

File: packages/stripe/subscriptions.ts
Comment (alannnc) [inline packages/stripe/subscriptions.ts:32]: We should be careful of not creating more than 1 sub per user

File: apps/web/components/ui/UsernameAvailability.tsx
Comment (leog) [reply apps/web/components/ui/UsernameAvailability.tsx:25]: I don't think so, it is within the same domain if I'm not mistaken

File: packages/app-store/stripepayment/api/subscription.ts
Comment (zomars) [reply packages/app-store/stripepayment/api/subscription.ts:83]: We should immediately charge if the user is upgrading only.

**Review by alannnc — COMMENTED** [review]
Thoughts on zomar's live review/discuss

- Validate self hosted users don't trigger stripe
- Move IntentUsername data in user metadata using hasKeyInMetadataFunction
- save requestedPremiumUsername/requestedUsername as metadata on stripe's end

**Comment by Jaibles** [comment]
Implementation looks perfect from the loom video 👌

**Comment by PeerRich** [comment]
loom looks great. waiting for @zomars and @leog for final review

File: apps/web/components/ui/UsernameAvailability.tsx
Comment (alannnc) [reply apps/web/components/ui/UsernameAvailability.tsx:279]: This is a particular case where my ActionButtons its being used twice on the view, while one its always hidden according to if the view is desktop or mobile. While doing this I can specifically access the one that runs on the desktop test

File: packages/app-store/stripepayment/api/subscription.ts
Comment (alannnc) [reply packages/app-store/stripepayment/api/subscription.ts:83]: always_invoice it is then

File: apps/web/components/ui/UsernameAvailability.tsx
Comment (zomars) [inline apps/web/components/ui/UsernameAvailability.tsx:18]: This will always fail on E2E tests on GitHub

**Review by zomars — CHANGES_REQUESTED** [review]
- E2E tests are failing locally and on GH even after the env were added
- There an infinite loop with `/api/username` now. Before this was forwarded to our `cal.com` website endpoint and now it calls itself.

**Comment by zomars** [comment]
@alannnc do you need help with this?

**Comment by zomars** [comment]
Made some fixes but got stuck here:
- Whenever we press Enter on the username field it triggers a updateProfile
  image
- We need to prevent it to submit the whole form

**Comment by PeerRich** [comment]
what's the state of this? @zomars can you follow up with @alannnc to get this over the finish line

File: packages/app-store/stripepayment/api/subscription.ts
Comment (zomars) [inline packages/app-store/stripepayment/api/subscription.ts:38]: [Prefer early returns](https://app.gitbook.com/o/6snd8PyPYMhg0wUw6CeQ/s/VXRprBTuMlihk37NQgUU/engineering/best-practices/prefer-early-returns)

```suggestion
  if (userCurrentPlan === UserPlan.PRO) {
    if (isNewUsernamePremium) return UsernameChangeStatusEnum.UPGRADE;
    return UsernameChangeStatusEnum.DOWNGRADE;
  }
  return UsernameChangeStatusEnum.NORMAL;
```

File: packages/app-store/stripepayment/api/subscription.ts
Comment (zomars) [inline packages/app-store/stripepayment/api/subscription.ts:42]: [Prefer defaultHandler for simple API handlers](https://app.gitbook.com/o/6snd8PyPYMhg0wUw6CeQ/s/VXRprBTuMlihk37NQgUU/engineering/best-practices/prefer-defaulthandler-for-simple-api-handlers)

File: apps/web/pages/api/intent-username/index.ts
Comment (zomars) [inline apps/web/pages/api/intent-username/index.ts:10]: 1. We're not checking for http method in this endpoint
2. [Prefer defaultHandler for simple API handlers](https://app.gitbook.com/o/6snd8PyPYMhg0wUw6CeQ/s/VXRprBTuMlihk37NQgUU/engineering/best-practices/prefer-defaulthandler-for-simple-api-handlers)

File: apps/web/pages/api/intent-username/index.ts
Comment (zomars) [inline apps/web/pages/api/intent-username/index.ts:27]: We can get `id` and `metadata` from first query instead of double fetching.

File: apps/web/pages/api/intent-username/index.ts
Comment (zomars) [inline apps/web/pages/api/intent-username/index.ts:32]: We can use zod for this. I think there's a global one on `packages/prisma/zod-utils.ts`

**Review by zomars — CHANGES_REQUESTED** [review]
Almost there!

File: apps/web/components/ui/UsernameAvailability/PremiumTextfield.tsx
Comment (alannnc) [reply apps/web/components/ui/UsernameAvailability/PremiumTextfield.tsx:187]: Woop its not, debugging.

**Review by alannnc — COMMENTED** [review]
@zomars I've tackled down changes suggested. I think we are ready to another revision. 🤙

**Review by zomars — APPROVED** [review]
Looking gucci 👌🏽 

__
[sema-logo](https://semasoftware.com/gh) &nbsp;**Summary:** :ok_hand: This code looks good

**Pull request merged**
